File: apps/api/src/routes/webhook.ts
import { Router, type Request, type Response, type Router as ExpressRouter } from "express";
import { Webhook } from "svix";
import { config } from "@/config/index.js";
import { um } from "@/lib/api-user-message.js";
import { sendJsonError } from "@/lib/http-json-response.js";
import { createLogger } from "@/utils/logger.js";
import { toLoggableError } from "@/utils/to-loggable-error.js";
import type { ClerkWebhookEvent } from "@/types/webhook/webhook.types.js";
import { processClerkWebhookDelivery } from "@/contexts/clerk-webhook-context.js";
import { rateLimitMiddleware } from "@/middleware/rate-limit.js";
import { toUserMessage, userFacingPayload } from "@/types/user-message.js";

const router: ExpressRouter = Router();
const logger = createLogger("routes:webhook");

router.post("/clerk", rateLimitMiddleware, async (req: Request, res: Response) => {
  try {
    const svixId = req.headers["svix-id"] as string;
    const svixTimestamp = req.headers["svix-timestamp"] as string;
    const svixSignature = req.headers["svix-signature"] as string;

    if (!svixId || !svixTimestamp || !svixSignature) {
      logger.warn(
        "Webhook request missing svix headers: %s, %s, %s",
        svixId,
        svixTimestamp,
        svixSignature || "undefined",
      );
      return sendJsonError(
        res,
        400,
        "Bad Request",
        um("MISSING_SVIX", "missingSvixHeaders", "Missing svix headers"),
      );
    }

    const payload = req.body instanceof Buffer ? req.body.toString() : String(req.body);

    const wh = new Webhook(config.clerkWebhookSecret);

    let evt: ClerkWebhookEvent;
    try {
      evt = wh.verify(payload, {
        "svix-id": svixId,
        "svix-timestamp": svixTimestamp,
        "svix-signature": svixSignature,
      }) as ClerkWebhookEvent;
    } catch (err: unknown) {
      logger.error(toLoggableError(err), "Webhook verification failed");
      return sendJsonError(
        res,
        400,
        "Bad Request",
        um("WEBHOOK_VERIFY_FAILED", "webhookVerificationFailed", "Webhook verification failed"),
      );
    }

    await processClerkWebhookDelivery(svixId, evt);

    return res.status(200).json({
      success: true,
      ...userFacingPayload(toUserMessage("WEBHOOK_OK", { key: "api.webhookProcessed" }, "Webhook processed")),
    });
  } catch (error: unknown) {
    logger.error(toLoggableError(error), "Error processing webhook");
    return sendJsonError(
      res,
      500,
      "Internal Server Error",
      um("FAILED_PROCESS_WEBHOOK", "failedProcessWebhook", "Failed to process webhook"),
    );
  }
});

export default router;

File: apps/api/src/internal/transport/http/webhook.go
package routes

import (
	"encoding/json"
	"io"
	"net/http"

	"github.com/labstack/echo/v4"
	svix "github.com/svix/svix-webhooks/go"

	"linky-api/src/internal/app/clerkwebhook"
	"linky-api/src/internal/config"
	"linky-api/src/internal/httpx"
	"linky-api/src/internal/logger"
)

var webhookLog = logger.New("routes:webhook")

func RegisterWebhook(g *echo.Group, cfg *config.Config) {
	g.POST("/clerk", func(c echo.Context) error {
		svixID := c.Request().Header.Get("svix-id")
		svixTs := c.Request().Header.Get("svix-timestamp")
		svixSig := c.Request().Header.Get("svix-signature")
		if svixID == "" || svixTs == "" || svixSig == "" {
			webhookLog.Warn().Msg("Webhook request missing svix headers")
			return httpx.SendError(c, http.StatusBadRequest, "Bad Request",
				httpx.UM("MISSING_SVIX", "missingSvixHeaders", "Missing svix headers"))
		}

		body, err := io.ReadAll(c.Request().Body)
		if err != nil {
			return httpx.SendError(c, http.StatusBadRequest, "Bad Request",
				httpx.UMDetail("WEBHOOK_VERIFY_FAILED", err.Error()))
		}

		if cfg.ClerkWebhookSecret != "" {
			wh, err := svix.NewWebhook(cfg.ClerkWebhookSecret)
			if err != nil {
				webhookLog.Error().Err(err).Msg("Webhook secret invalid")
				return httpx.SendError(c, http.StatusInternalServerError, "Internal Server Error",
					httpx.UM("FAILED_PROCESS_WEBHOOK", "failedProcessWebhook", "Failed to process webhook"))
			}
			if err := wh.Verify(body, c.Request().Header); err != nil {
				webhookLog.Error().Err(err).Msg("Webhook verification failed")
				return httpx.SendError(c, http.StatusBadRequest, "Bad Request",
					httpx.UM("WEBHOOK_VERIFY_FAILED", "webhookVerificationFailed", "Webhook verification failed"))
			}
		}

		var evt map[string]interface{}
		if err := json.Unmarshal(body, &evt); err != nil {
			return httpx.SendError(c, http.StatusBadRequest, "Bad Request",
				httpx.UMDetail("WEBHOOK_VERIFY_FAILED", err.Error()))
		}

		if err := clerkwebhook.ProcessDelivery(c.Request().Context(), svixID, evt); err != nil {
			webhookLog.Error().Err(err).Msg("Error processing webhook")
			return httpx.SendError(c, http.StatusInternalServerError, "Internal Server Error",
				httpx.UM("FAILED_PROCESS_WEBHOOK", "failedProcessWebhook", "Failed to process webhook"))
		}

		fallback := "Webhook processed"
		um := httpx.UserMessage{
			Code:            "WEBHOOK_OK",
			I18n:            &httpx.I18nPayload{Key: "api.webhookProcessed"},
			FallbackMessage: &fallback,
		}
		return httpx.SendUserMessage(c, http.StatusOK, map[string]interface{}{"success": true}, um)
	})
}
